### PR TITLE
[SYCL] kernel_compiler device flag fix

### PR DESCRIPTION
**Pull request by cperkinsintel**
We want to ensure -Xs device flags are passed through to device compiler when compiling with the kernel_compiler.

### CODE REVIEW

File: sycl/test-e2e/KernelCompiler/sycl_device_flags.cpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// REQUIRES: level_zero
+// UNSUPPORTED: windows
+
+// IGC shader dump not available on Windows
```
Comment (maarquitos14) [inline sycl/test-e2e/KernelCompiler/sycl_device_flags.cpp:12]:
```suggestion
// IGC shader dump not available on Windows.
```

File: sycl/test-e2e/KernelCompiler/sycl_device_flags.cpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+// RUN: %{build} -o %t.out
+// RUN: env IGC_DumpToCustomDir=%t.dump IGC_ShaderDumpEnable=1 NEO_CACHE_PERSISTENT=0 %{run} %t.out
+// RUN: grep -e '-doubleGRF' %t.dump/OCL_asmaf99e2d4667ef6d3_options.txt
```
[the remark this answers is not in the document]
Comment (cperkinsintel) [reply sycl/test-e2e/KernelCompiler/sycl_device_flags.cpp:16]: I had thought the same, but it seems to not be random, probably a hash of some sort

File: sycl/source/detail/kernel_bundle_impl.hpp [hunk 3]
```diff
@@ -360,6 +360,43 @@ class kernel_bundle_impl {
...
+    return "";
+  }
+
+  std::string extractXsFlags(const std::vector<std::string> &BuildOptions) {
+    std::stringstream SS;
+    for (std::string Option : BuildOptions) {
+      auto Where = Option.find("-Xs ");
+      if (Where != std::string::npos) {
+        Where += 4;
+        std::string flags = Option.substr(Where);
```
[the remark this answers is not in the document]
Comment (cperkinsintel) [reply sycl/source/detail/kernel_bundle_impl.hpp:393]: ?

File: sycl/source/detail/kernel_bundle_impl.hpp [hunk 4]
```diff
@@ -360,6 +360,43 @@ class kernel_bundle_impl {
...
+      return std::string(start, end);
+    }
+    // else
+    return "";
+  }
+
+  std::string extractXsFlags(const std::vector<std::string> &BuildOptions) {
+    std::stringstream SS;
+    for (std::string Option : BuildOptions) {
+      auto Where = Option.find("-Xs ");
```
[the remark this answers is not in the document]
Comment (cperkinsintel) [reply sycl/source/detail/kernel_bundle_impl.hpp]: It was intentional, but wrong. `-Xs` should work without spaces. I've updated the code and the test.  Thanks for catching this!

File: sycl/source/detail/kernel_bundle_impl.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (maarquitos14) [reply sycl/source/detail/kernel_bundle_impl.hpp:393]: Sorry, meant to do this:
```suggestion
        std::string Flags = Option.substr(Where);
```

**Review by maarquitos14 — APPROVED** [review]
LGTM.

File: sycl/source/detail/kernel_compiler/kernel_compiler_sycl.cpp [hunk 5]
```diff
@@ -165,61 +162,71 @@ std::string getCompilerName() {
...
+#ifdef _WIN32
+  FILE *pipe = _popen(command.c_str(), "r");
+#else
+  FILE *pipe = popen(command.c_str(), "r");
+#endif
+  if (!pipe) {
+    return -1;
+  }
+
+  char buffer[124];
```
Comment (maarquitos14) [inline sycl/source/detail/kernel_compiler/kernel_compiler_sycl.cpp:175]: It's just a nit, but 124 is a weird size, any reason to choose it?

**Review by maarquitos14 — APPROVED** [review]
Just a nit, otherwise LGTM.

File: sycl/source/detail/kernel_compiler/kernel_compiler_sycl.cpp (hunk 5 above)
Comment (cperkinsintel) [reply sycl/source/detail/kernel_compiler/kernel_compiler_sycl.cpp:175]: Initially I had it at 1024. I must have deleted that 0 at some point. I have more kernel_compiler work upcoming, I'll restore it there.

**Comment by sarnex** [comment]
@cperkinsintel Seeing some postcommit fails:

```
 ********************
Failed Tests (2):
  SYCL :: KernelCompiler/kernel_compiler_sycl.cpp
  SYCL :: KernelCompiler/sycl_device_flags.cpp

```


Can you please take a look? Thanks

**Comment by againull** [comment]
> @cperkinsintel Seeing some postcommit fails:
> 
> ```
>  ********************
> Failed Tests (2):
>   SYCL :: KernelCompiler/kernel_compiler_sycl.cpp
>   SYCL :: KernelCompiler/sycl_device_flags.cpp
> ```
> 
> https://github.com/intel/llvm/actions/runs/10079697084/job/27868764383
> 
> https://github.com/intel/llvm/actions/runs/10079697084/job/27868764957
> 
> Can you please take a look? Thanks

Chris opened the following PR with the fix: https://github.com/intel/llvm/pull/14752
I forgot to post it here.

**Comment by sarnex** [comment]
ah ok thanks, missed it :)

**Pull request merged**